Ensure set object descriptions are reproducible.

lamby opened the pull request:
Whilst working on the Reproducible Builds effort [0], we noticed
that sphinx could generate output that is not reproducible.

In particular, the rendering of `set` objects in default arguments
and elsewhere is currently non-determinstic. For example:

    class A_Class(object):
        a_set = {'a', 'b', 'c'}

Might be rendered as any of:

     {'a', 'b', 'c'}
     {'a', 'c', 'b'}
     {'b', 'a', 'c'}
     {'b', 'c', 'a'}
     {'c', 'a', 'b'}
     {'c', 'b', 'a'}

Patch attached that sorts the contents of sets whilst rendering.
This is parallel to the `dict` key sorting.

This was originally filed in Debian as #895553 [1].

 [0] https://reproducible-builds.org/
 [1] https://bugs.debian.org/895553

tk0miya commented on sphinx/util/inspect.py:
```diff
@@ -246,6 +246,10 @@ def object_description(object):
         else:
             items = ("%r: %r" % (key, object[key]) for key in sorted_keys)
```
I think values should be also processed for dicts contains set as a value.

mitya57 replied:
How about calling `object_description` recursively? And below too, replace`repr(x)` with `object_description(x)`.

tk0miya replied:
Yes, `object_description(object[key])` is better to me.

mitya57 replied:
Ok. This was my fault when I added the dict processing code, but I hope @lamby will fix it :)

lamby replied:
Pushed, including the call to `object_description` in the dict handling too :)

lamby committed:
Call object_description recursively for dict keys and values.

tk0miya commented on sphinx/util/inspect.py:
```diff
@@ -244,8 +244,15 @@ def object_description(object):
+        # Sort set contents
+        template = "{%s}" if PY3 else "set([%s])"
+        return template % ", ".join(object_description(x)
+                                    for x in sorted(object))
```
This will be failed if members are not sortable:
```
>>> sorted({1, 2, 'str'})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'str' and 'int'
```
```
>>> sorted({1, 2, object()})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'object' and 'int'
```

mitya57 replied:
So a `try…except` is needed, like in the dict case.

lamby replied:
Good shout. Have updated with a testcase.

lamby committed:
Ensure the set object description is reproducible.

Whilst working on the Reproducible Builds effort [0], we noticed
that sphinx could generate output that is not reproducible.

In particular, the rendering of `set` objects in default arguments
and elsewhere is currently non-determinstic. For example:

    class A_Class(object):
        a_set = {'a', 'b', 'c'}

Might be rendered as any of:

     {'a', 'b', 'c'}
     {'a', 'c', 'b'}
     {'b', 'a', 'c'}
     {'b', 'c', 'a'}
     {'c', 'a', 'b'}
     {'c', 'b', 'a'}

Patch attached that sorts the contents of sets whilst rendering.
This is parallel to the `dict` key sorting.

This was originally filed in Debian as #895553 [1].

 [0] https://reproducible-builds.org/
 [1] https://bugs.debian.org/895553

Signed-off-by: Chris Lamb <lamby@debian.org>

tk0miya reviewed (approved):
LGTM! I'll merge this soon.

lamby commented:
Thanks all!
